### ISSUE TITLE
"path too long" when cloning a repo

### ISSUE THREAD

**Issue by A-Ovchinnikov-mx** [libgit2/libgit2#6070]
### Reproduction steps

`git_clone` a repo. The repo has a path that, combined with a local checkout path, would be close enough to MAX_PATH.

See repo https://github.com/A-Ovchinnikov-mx/libgit2_bug_test for an example.

### Expected behavior

The repo is cloned successfully.

### Actual behavior

```
Failed to clone repo: path too long: 'C:/WORK/test/libgit2_bug_test/C:/WORK/test/libgit2_bug_test/a_very_long_path_that_almost_hits_the_limit_but_is_not_quiet_there/it_could_be_a_nodemodules_directory_for_example/that_contains_a_lot_of_nested_directories/combined_with_a_longer_path_to_the_repository.txt'
```

### Version of libgit2 (release number or SHA1)

v1.2.0 and 136901086ecfdd2b5cc106782310355a9c0b1a9a

### Operating system(s) tested

Windows 10 x64

### Additional information

One can notice that the path that is considered too long is actually wrong:

C:/WORK/test/libgit2_bug_test/**C:/WORK/test/libgit2_bug_test**

Without this, the path is 237 characters long which is not the limit. But with the checkout directory being added twice, this path becomes too long.

**Comment by A-Ovchinnikov-mx** [comment]
@ethomson Could you please postpone v1.3.0 release until this is fixed?

**Comment by A-Ovchinnikov-mx** [comment]
The issue is in `git_buf_joinpath` where:

```
a = C:/WORK/test/libgit2_bug_test/
b = C:/WORK/test/libgit2_bug_test/a_very_long_path_that_almost_hits_the_limit_but_is_not_quiet_there/it_could_be_a_nodemodules_directory_for_example/that_contains_a_lot_of_nested_directories/combined_with_a_longer_path_to_the_repository.txt
```

**Comment by A-Ovchinnikov-mx** [comment]
@ethomson I think the deletion of `git_path_join_n` caused this: https://github.com/libgit2/libgit2/commit/97769280ba9938ae27f6e06cbd0d5e8a768a86b9#diff-d534511b316cb989d76fcf2dee0545e386f9b316bc911519ca9bc01fc923ca6aL191

Now it's a simple `git_buf_join` that doesn't do much path calculation.

**Comment by ethomson** [comment]
This is because `collect_attr_files` prefixes the workdir path to an already absolute path.  This went unnoticed until now because we didn't enforce path lengths ourselves, we passed it on to Windows and just let it return a not found error.

**Comment by ethomson** [comment]
There's two things that we need to do here:

1. Prefixing the workdir to an already-absolute path is obviously wrong.  We'll need to fix this.  Unfortunately, I have no idea how prevalent this is.
2. We might want to consider changing longpath validation so that there's a fail mode or not.  It might make sense for longpath validation to act as if a file were not found when reading but actually fail when trying to write.

**Comment by A-Ovchinnikov-mx** [comment]
> This went unnoticed until now because we didn't enforce path lengths ourselves, we passed it on to Windows and just let it return a not found error.

@ethomson I'm not sure what do you mean tbh but v1.1.1 doesn't try to do that so it works correctly 🙂 

Will you be able to fix that before v1.3.0 release?

**Comment by A-Ovchinnikov-mx** [comment]
This seems to fix it; not sure if this is a good fix or not:

```diff
diff --git a/src/attr.c b/src/attr.c
index 03b720c5a..5edb9e869 100644
--- a/src/attr.c
+++ b/src/attr.c
@@ -621,8 +621,13 @@ static int collect_attr_files(
 
 	/* Resolve path in a non-bare repo */
 	if (workdir != NULL) {
-		if (!(error = git_repository_workdir_path(&dir, repo, path)))
-			error = git_path_find_dir(&dir);
+	    if (git_path_is_absolute(path))
+		error = git_buf_puts(&dir, path);
+	    else
+		error = git_repository_workdir_path(&dir, repo, path);
+
+	    if (!error)
+		error = git_path_find_dir(&dir);
 	}
 	else {
 		error = git_path_dirname_r(&dir, path);
```

**Comment by ethomson** [comment]
> @ethomson I'm not sure what do you mean tbh but v1.1.1 doesn't try to do that so it works correctly 🙂

That's because we added `core.longpaths` in v1.2.0.  We had the same bug where we were looking in the wrong path, we just didn't notice it because we weren't doing the long paths validation ourselves.

> This seems to fix it; not sure if this is a good fix or not:

Maybe so!  I would like to do a little more research.  I'm not sure why we're passing a relative path sometimes and not others.  I might like to split the API into two very clear ones so that this is more obvious.

> Will you be able to fix that before v1.3.0 release?

Need to do a little more investigation, but this is pretty nasty, so I would expect so.

**Comment by ethomson** [comment]
@A-Ovchinnikov-mx can you see if #6073 solves your problem?

**Comment by ethomson** [comment]
Also #6075 

I did a pass on the places where we validate workdir paths for length.  I don't immediately see any places where we should be treating it as `ENOTFOUND` instead of just `-1`, but I suspect that there are some.  For example, you can imagine that we construct a path to a `.gitignore` that is slightly longer than the max path length, but it is in a folder that has files like `a`, `b`, `c`.  In theory, this should be permissible, but you're basically at the knife's edge of max path length anyway, so I'm not too overly concerned if it fails in a slightly different way than you expect.

(One could argue that we _should_ fail if any git configuration path would be too long.)

Anyway, these sort of edge cases are why we added `core.longpath` support in the first place.

**Comment by A-Ovchinnikov-mx** [comment]
> @A-Ovchinnikov-mx can you see if #6073 solves your problem?

This seems to resolve the issue indeed, at least I can now clone the test repo.

**Comment by A-Ovchinnikov-mx** [comment]
> Also #6075

This one alone doesn't solve the issue but I don't expect it to looking at the changes 🙂

**Comment by ethomson** [comment]
Fixed via #6073
